feat: add fastrace blog (#1)
fast/fast.github.io: src/content/post/fastrace-a-modern-approach-to-distributed-tracing-in-rust.mdx

---
title: "Fastrace: A Modern Approach to Distributed Tracing in Rust"
tags:
  - Observability
  - Traces
pubDate: 2025-03-21
---

## TL;DR

> Distributed tracing is critical for understanding modern microservice architectures. While [`tokio-rs/tracing`](https://github.com/tokio-rs/tracing) has been the standard in Rust, it comes with significant challenges: ecosystem fragmentation, complex configuration, and high overhead. 

Reviewer: ```suggestion
Distributed tracing is critical for understanding modern microservice architectures. While `tokio-rs/tracing` is widely used in Rust, it comes with significant challenges: ecosystem fragmentation, complex configuration, and high overhead. 
```

> [`fastrace`](https://github.com/fast/fastrace) provides a production-ready solution with seamless ecosystem integration, out-of-box OpenTelemetry support, and a simpler API that works naturally with the existing logging infrastructure.

Reviewer: ```suggestion
[`fastrace`](https://github.com/fast/fastrace) provides a production-ready solution with seamless ecosystem integration, out-of-box OpenTelemetry support, and a more straightforward API that works naturally with the existing logging infrastructure.
```

> The following example demonstrates the simplicity of tracing functions with `fastrace`:

Reviewer: ```suggestion
The following example demonstrates how to trace functions with `fastrace`:
```

```rust
#[fastrace::trace]
pub fn send_request(req: HttpRequest) -> Result<(), Error> {
    // ...
}
```

> It's being used in production at companies like [ScopeDB](https://www.scopedb.io/), where it helps trace and debug petabyte-scale observability data workloads.

Reviewer: ```suggestion
It's being used in production by products like [ScopeDB](https://www.scopedb.io/blog/manage-observability-data-in-petabytes), where it helps trace and debug petabyte-scale observability data workloads.
```

![Distributed Tracing Visualization](assets/scopedb-traces.png)

## Why Distributed Tracing Matters

> In today's world of microservices and distributed systems, understanding what's happening inside your applications has never been more challenging. When a single user request might touch dozens of services before completion, traditional logging approaches quickly fall short.

Reviewer: ```suggestion
Understanding what is happening inside your applications has never been more challenging in today's microservices and distributed systems. A user request might touch dozens of services before completion, and traditional logging approaches quickly fall short.
```

Move the real challenge in front.

Consider a typical request flow:

```
User → API Gateway → Auth Service → User Service → Database
```

> When something goes wrong or performs poorly, where exactly is the issue? Individual service logs only show fragments of the story, lacking the crucial context of how the request flows through your entire system.

Reviewer: ```suggestion
When an exception is thrown, or the app performs poorly, where exactly is the root cause? Individual service logs only show fragments of the whole trace, lacking the crucial context of how the request flows through your entire system.
```

> This is where distributed tracing becomes essential. Tracing creates a connected view of your request's journey across service boundaries, making it possible to:

Reviewer: ```suggestion
This is where distributed tracing becomes essential. Tracing creates a connected view of your request's flow across service boundaries, making it possible to:
```

Keep wording in sync.

- Identify performance bottlenecks across services
- Debug complex interactions between components
- Understand dependencies and service relationships
- Analyze latency distributions and outliers
- Correlate logs and metrics with request context

## A Famous Approach: `tokio-rs/tracing`

> For many Rust developers, the `tokio-rs/tracing` ecosystem has been the go-to solution for application instrumentation. Let's look at how a typical implementation works:

Reviewer: ```suggestion
For some Rust developers, `tokio-rs/tracing` is the go-to solution for application instrumentation. Let's look at how a typical implementation works:
```

```rust
fn main() {
    // Initialize the tracing subscriber
    // Complex configuration code omitted...
    
    // Create a span and record some data
    let span = tracing::info_span!("processing_request", 
        user_id = 42, 
        request_id = "abcd1234"
    );
    
    // Enter the span (activates it for the current execution context)
    let _guard = span.enter();
    
    // Log within the span context
    tracing::info!("Starting request processing");
    
    process_data();
    
    tracing::info!("Finished processing request");
}
```

For instrumenting functions, `tokio-rs/tracing` provides attribute macros:

```rust
use tracing::instrument;

#[instrument(skip(password), fields(user_id = user.id))]
async fn authenticate(user: &User, password: &str) -> Result<AuthToken, AuthError> {
    info!("Authenticating user {}", user.id);
    // ...more code...

> }

Reviewer: ```suggestion
#[tracing::instrument(skip(password), fields(user_id = user.id))]
async fn authenticate(user: &User, password: &str) -> Result<AuthToken, AuthError> {
    tracing::info!("Authenticating user {}", user.id);
    // ...more code...
}
```

```

## The Challenges with `tokio-rs/tracing`

> While powerful, `tokio-rs/tracing` comes with several significant challenges:

Reviewer: ```suggestion
According to our previous user experience, `tokio-rs/tracing` comes with several significant challenges:
```

### 1. Ecosystem Fragmentation

By introducing its own logging macros, `tokio-rs/tracing` creates a division with code using the standard `log` crate:

```rust
// Using log crate
log::info!("Starting operation");

// Using tracing crate (different syntax)
tracing::info!("Starting operation");
```

This fragmentation is particularly problematic for library authors. When creating a library, authors face a difficult choice:

1. Use the `log` crate for compatibility with the broader ecosystem
2. Use `tokio-rs/tracing` for better observability features
3. Implement dual support for both systems (increasing maintenance burden)

    This often leads to complex feature-flag systems and wrapper macros:

    ```rust
    macro_rules! info {
        ($($arg:tt)*) => {
            if cfg!(feature = "tracing") {
                tracing::info!($($arg)*);
            } else {
                log::info!($($arg)*);
            }
        }
    }

>     ```

Reviewer: One may argue that tracing has a `"log"` feature flag so they can always use tracing with optionally `log` flag enabled.

Author: refined

> # In a library's Cargo.toml
> [dependencies]
> tracing = { version = "0.1", features = ["log"] }  # Required for log compatibility
> 
> # In an application's Cargo.toml
> [dependencies]
> tracing = "0.1"
> tracing-log = "0.2"  # Required bridge for log crate compatibility

Reviewer: ```suggestion
# Library's Cargo.toml
[dependencies]
tracing = { version = "0.1", features = ["log"] }  # Emit a log record for log compatibility

# Application's Cargo.toml
[dependencies]
tracing = "0.1"
tracing-log = "0.2"  # Listen to log records for log compatibility
```

Many libraries choose the first option for simplicity, but miss out on the benefits of tracing. This creates a situation where applications have inconsistent instrumentation across their dependency tree.

### 2. Performance Impact for Libraries

Library authors are particularly sensitive to performance overhead, as their code may be called in tight loops or performance-critical paths. `tokio-rs/tracing`'s overhead can be substantial when enabled, which creates a dilemma:

1. Always enable tracing (and impose overhead on all users)
2. Don't instrument at all (and lose observability)
3. Create a complex feature-flag system (increasing maintenance burden)

    This is a common pattern in libraries using `tokio-rs/tracing`:

    ```rust
    #[cfg_attr(feature = "tracing", tracing::instrument(skip(password), fields(user_id = user.id)))]
    async fn authenticate(user: &User, password: &str) -> Result<AuthToken, AuthError> {
        // ...more code...
    }
    ```

> With `tokio-rs/tracing`, there's no clean way to have zero-cost tracing when disabled. This makes library authors reluctant to add instrumentation to performance-sensitive code paths.

Reviewer: fastrace has an `enabled` flag. But tracing doesn't have any similar mechanism?

I don't get it quite clearly what "no clean way to have zero-cost tracing when disabled" means, especially what "when (tracing) disabled" means, or how to.

>     ```

Reviewer: ````suggestion
3. Create an additional feature flag system (increasing maintenance burden)

Here is a common pattern in libraries using `tokio-rs/tracing`:

```rust
#[cfg_attr(feature = "tracing", tracing::instrument(skip(password), fields(user_id = user.id)))]
async fn authenticate(user: &User, password: &str) -> Result<AuthToken, AuthError> {
        // ...more code...
}
```

Different libraries may define flags with subtly different names, making it hard for the final application to configure all of them.
````

### 3. No Context Propagation

Distributed tracing requires propagating context across service boundaries, but `tokio-rs/tracing` leaves this largely as an exercise for the developer. For example, the [tonic official example](https://github.com/hyperium/tonic/blob/master/examples/src/tracing/server.rs) demonstrates how to trace gRPC service using tonic:

```rust

> // Server setup with `tokio-rs/tracing`

Reviewer: ```suggestion
```

For tidy.

Server::builder()
    // Basic span only, no context extraction
    .trace_fn(|_| tracing::info_span!("grpc_server"))
    .add_service(MyServiceServer::new(MyService::default()))
    .serve(addr)
    .await?;
```

> The above example only creates a basic span but doesn't extract tracing context from the incoming request.

Reviewer: Can we elaborate a bit here how the configure above would cause a (maybe?) fragmented trace?

As a reader one may think that: so, you can configure tracing in one line, what's the problem here?

Author: refined with additional examples

> Distributed tracing requires propagating context across service boundaries, but `tokio-rs/tracing` leaves this largely as an exercise for the developer. For example, this is the [tonic official example](https://github.com/hyperium/tonic/blob/master/examples/src/tracing/server.rs) demonstrating how to trace a gRPC service:

Reviewer: ```suggestion
Distributed tracing requires propagating context across service boundaries, but `tokio-rs/tracing` leaves this largely as an exercise for the developer. For example, this is [tonic's official example](https://github.com/hyperium/tonic/blob/master/examples/src/tracing/server.rs) demonstrating how to trace a gRPC service:
```

nit syntax

> ## Introducing `fastrace`: A Modern Solution

Reviewer: ```suggestion
## Introducing `fastrace`: A Fast and Complete Solution
```

Fastrace provides a better solution, not only by it's a new one.

`fastrace` takes a different approach to tracing in Rust. Its core design principles include:

### 1. Zero-cost Abstraction

> `fastrace` is designed as a true zero-cost abstraction. When disabled, instrumentation compiles away completely, resulting in no runtime overhead. This makes it ideal for libraries concerned about performance.

Reviewer: ```suggestion
`fastrace` is designed with real zero-cost abstraction. When disabled, instrumentations are completely omitted from compilation, resulting in no runtime overhead. This makes it ideal for libraries concerned about performance.
```

### 2. Ecosystem Compatibility

`fastrace` focuses exclusively on distributed tracing. Through its composable design, it integrates seamlessly with the existing Rust ecosystem, including compatibility with the standard `log` crate. This architectural approach allows libraries to implement comprehensive tracing while preserving their users' freedom to use their preferred logging setup.

### 3. Simplicity First

The API is designed to be intuitive and require minimal boilerplate, focusing on the most common use cases while still providing extensibility when needed.

### 4. Insanely Fast

![Fastrace Performance](assets/trace-100-spans.png)

> `fastrace` is designed for high-performance applications. It can handle massive amount of spans per second with minimal impact on CPU and memory usage.

Reviewer: ```suggestion
`fastrace` is designed for high-performance applications. It can handle massive amounts of spans with minimal impact on CPU and memory usage.
```

Here is no concrete number so "per second" is wordy.

}
```

The key point here is that libraries should include fastrace without enabling any features:

```toml
[dependencies]
fastrace = "0.7"  # No 'enable' feature
```

When an application uses this library and doesn't enable the 'enable' feature of fastrace:
- All tracing code is completely optimized away at compile time
- Zero runtime overhead is added to the library
- No impact on performance-critical code paths

When the application does enable tracing via the 'enable' feature:
- Instrumentation in the dedicated library becomes active
- Spans are collected and reported
- The application gets full visibility into library behavior

This is a significant advantage over other tracing solutions that either always impose overhead or require libraries to implement complex feature-flag systems.

### 6. Seamless Context Propagation

`fastrace` provides companion crates for popular frameworks that handle context propagation automatically:

```rust
// For HTTP clients with reqwest
let response = client.get(&format!("https://user-service/users/{}", user_id))
    .headers(fastrace_reqwest::traceparent_headers())
    .send()
    .await?;

// For gRPC servers with tonic
Server::builder()

>     .layer(FastraceServerLayer)  // Automatically extracts context from incoming requests

Reviewer: ```suggestion
    .layer(fastrace_tonic::FastraceServerLayer)  // Automatically extracts context from incoming requests
```

I guess it's from`fastrace_tonic`? If so, let's make it clear. Ditto the following bridge.

    .add_service(MyServiceServer::new(MyService::default()))
    .serve(addr)
    .await?;

// For gRPC clients
let channel = ServiceBuilder::new()
    .layer(FastraceClientLayer)  // Automatically injects context into outgoing requests
    .service(channel);

// For data access with OpenDAL
let op = Operator::new(services::Memory::default())?
    .layer(FastraceLayer)  // Automatically traces all data operations
    .finish();
op.write("test", "0".repeat(16 * 1024 * 1024).into_bytes())
    .await?;
```

This provides out-of-box distributed tracing without manual context handling.

## The Complete Solution: `fastrace` + `log` + `logforth`

`fastrace` deliberately focuses on doing one thing well: tracing. Through its composable design and the Rust's great ecosystem, a powerful combination emerges:

- **log**: The standard Rust logging facade
- **logforth**: A flexible logging implementation with industrial-ready features
- **fastrace**: High-performance tracing with distributed context propagation

> The integration allows your logs to be automatically associated with trace spans, providing correlation without requiring new logging APIs:

Reviewer: ```suggestion
This integration allows automatically associating your logs with trace spans, providing correlation without requiring using different logging macros:
```

> // For data access with OpenDAL

Reviewer: ```suggestion
// For data access with Apache OpenDAL
```

nit

```rust
// Standard log entries are automatically associated with the current span
log::info!("Processing started");

// Later, in your logging infrastructure, you can see which trace and span
// each log entry belongs to
```

To illustrate the simplicity of this approach, here's a streamlined example of building a microservice with complete observability:

```rust
use fastrace::prelude::*;

> use fastrace_poem::FastraceMiddleware;
> use fastrace_jaeger::JaegerReporter;
> use logforth::diagnostic;

Reviewer: ```suggestion
```

Perhaps inline all the FQN. I can see that `logforth::diagnostic` is used full-qualified below.

---
title: "Fastrace: A Modern Approach to Distributed Tracing in Rust"
tags:
  - Observability
  - Traces
pubDate: 2025-03-21
---

## TL;DR

Distributed tracing is critical for understanding modern microservice architectures. While `tokio-rs/tracing` is widely used in Rust, it comes with significant challenges: ecosystem fragmentation, complex configuration, and high overhead. 

> [fastrace](https://github.com/fast/fastrace) provides a production-ready solution with seamless ecosystem integration, out-of-box OpenTelemetry support, and a more straightforward API that works naturally with the existing logging infrastructure.

Reviewer: ```suggestion
[Fastrace](https://github.com/fast/fastrace) provides a production-ready solution with seamless ecosystem integration, out-of-box OpenTelemetry support, and a more straightforward API that works naturally with the existing logging infrastructure.
```

nit
